text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+};
+
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+     semverGT(resource.apiVersion, semver(2,1,0))
+};
+```
+
+For simplicity, safety, and readability, Cedar functions cannot call other functions, and cannot take functions as arguments.
+
+## Motivation
+
+Cedar currently lacks mechanisms for users to build abstractions. The only existing mechanism is extension functions, which are insufficient for two reasons:
+
+1. They are extremely heavyweight, requiring modifying the source of the Cedar evaluator. This means that users who want to stay on official versions of Cedar have no choice but to attempt to submit a PR and get it accepted into the mainline. This process does not scale.
+    1. For data structures that are relatively standard (ex: SemVer, or OID Users as proposed in [RFC 58](https://github.com/cedar-policy/rfcs/blob/cdisselkoen/standard-library/text/0058-standard-library.md)), it’s hard to know what’s in-demand enough to be included, and how to balance that against avoiding bloat. There’s no way to naturally observe usage because the only way to “install” the extension pre-acceptance is to vend a modified version of Cedar.
+    2. Users may have data structures that are totally bespoke to their systems. It makes no sense to include these in the standard Cedar distribution at all, yet users may still want some way to build abstractions. 
+2. They are too powerful. Extensions are implemented via arbitrary Rust code, which is essential for encoding features that cannot be represented via Cedar expressions (such as IP Addresses), but opens the door for a wide range of bugs/design issues. It’s trivial to design an extension that is difficult to validate and/or logically encode for analysis. Problematically, extension functions can potentially exhibit non-determinism, non-termination, or non-linear performance; interact with the operating system; or violate memory safety. This raises the code review burden when considering an extension function's implementation.
+
+In contrast, functions written as simple abstractions over Cedar expressions, which themselves cannot call other Cedar functions, have none of these problems. They naturally inherit the properties of Cedar expressions. Cedar functions are guaranteed to terminate and be deterministic. Since they are compositions of Cedar expressions, it’s easy to validate and analyze them.
+
+## Detailed Design
+
+### Function declarations
+
+This RFC adds a new top level form to Cedar policysets: the function declaration.
+
+A Cedar function declaration is composed of three elements:
+
+1. A name, which is a valid (possibly namespaced) identifier.
+2. A list of parameters, each of which is a non-namespaced identifier.
+3. A body, which is a Cedar expression.
+    1. The body may contain (non-namespaced) variables, drawn from the function's parameters.
+    2. The body may not contain calls to other functions.
```

Comment:
They _can_ contain these variables now that we have renamed function parameters to start with `?`.

And: I don't think we should call these functions, we should call them macros. More on this in response to a later comment.

Reply:
I still don't think they should, violates macro hygiene and could lead to confusion

Reply:
It doesn't violate macro hygeine. `principal` etc. are not macro variables.

Reply:
But your referencing a variable not bound in the macro body

Reply:
From Hygienic Macro Expansion
> Hygiene Condition for Macro Examples:
Generated identifiers that become binding instances in the completely expanded program must only bind variables that are generated at the same transcription step.

In this: bindings generated in the first phase (the syntax of policies) are bound in the second phase (macro-expansion) 

Why we care:
It's backwards compatible to add the ability to directly reference principal/action/resource. If we end up seeing examples that would be a lot clearer with this we can add it. On the other hand, if we expand the language at any point to include other binding forms, having strict hygiene will become much more important, and it will be backwards incompatible to remove direct PAR references.

Reply:
I don't think I agree that this is a violation of hygiene. These are completely different sorts of variables, and they are global. There is no risk of confusion/capture/etc. of variables bound/applied in the macro.

But maybe it's an issue of clarity generally. Would we be Ok with
```
def isOwner() principal == resource.owner;
permit(principal,action,resource) when { isOwner() };
```
or do we really want
```
def isOwner(?principal,?resource) ?principal == ?resource.owner;
permit(principal,action,resource) when { isOwner(principal,resource) };
```
I can imagine preferring the first example so that you cannot mistakenly swap the arguments, e.g.,
```
def isOwner(?principal,?resource) ?principal == ?resource.owner;
permit(principal,action,resource) when { isOwner(resource,principal) };
```

Reply:
Hmm, I prefer the first one, but I've been thinking about it and I can't come up with a good reason why, so maybe it's just taste.

Reply:
I'm lukewarm on it, to be honest. It's more future-proof to forbid the use of `principal`, `action`, etc. within macro bodies, so that might be the best way to proceed.

It occurs to me that you might want to add a section about interaction with templates. Suggested text:

Cedar macros use the same notation for their variables as do templates, since in both cases variables are "holes" that are filled in with a Cedar expression to make a full construct -- for templates, the construct is a policy, for macros it is an expression.

Templates can include calls to macros. For example:
```
def semver(?major, ?minor, ?patch) ... // same as initial example
def semverGT(?lhs, ?rhs) ... // same as initial example
permit (
  principal == ?principal,
  action == Action::"view",
  resource in ?resource)
when {
  semverGT(resource.apiVersion, semver(2,1,0))
};
```
When we link this template, we get a full policy, as usual.

A macro parameter can end up having the same name as a template variable, with the effect that it shadows it, which is expected with normal lexical scoping: 
```
def isOwner(?principal,?resource) ?principal == ?resource.owner;
permit(principal,action,resource in ?resource) 
when { isOwner(principal,resource) };
```
Here, notice that the `isOwner` macro's reference to `?resource` is to its parameter, not to the template slot. (We'd recommend to users to choose different parameter names to avoid confusion.)

As already mentioned, you cannot refer to a parameter in a macro that it does not bind, which enforces hygiene. So the following is not allowed.
```
def isOwner() ?principal == ?resource.owner; 
// ERROR: cannot refer to unbound variables ?principal, ?resource
```